twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/WorkspaceChannelList.tsx

```diff
@@ -1,57 +1,90 @@
-  const collectionPath = `/channels/v1/companies/${companyId}/workspaces/${workspaceId}/channels/`;
-  const channelsCollection = Collection.get(collectionPath, ChannelResource);
-  const channels = channelsCollection.useWatcher({}, { limit: limit });
+  const currentUserId: string = UsersService.getCurrentUserId();
+  const inputRef = useRef<Input>(null);
 
   useEffect(() => {
-    channelsCollection.reload();
+    listService.bind(search);
   });
 
-  const minePath = `/channels/v1/companies/${companyId}/workspaces/${workspaceId}/channels/::mine`;
-  const mineCollection = Collection.get(minePath, ChannelResource);
-  const mine = mineCollection.useWatcher({});
+  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
+    if (event.key === 'ArrowUp' && cursor > 0) {
+      return setCursor(cursor - 1);
+    }
 
-  channels.map((channel: ChannelResource) => {
-    workspaceChannels.push({
-      id: channel.data.id || '',
-      name: channel.data.name || '',
-      type: 'workspace',
-      channelResource: channel,
-    });
-  });
+    if (event.key === 'ArrowDown' && cursor < listService.list.length - 1) {
+      if (cursor < limit - 1) return setCursor(cursor + 1);
+    }
+
+    if (event.key === 'Enter' && cursor >= 0) {
+      const element = listService.list[cursor];
+
+      return handleElementType(element);
+    }
+  };
+
+  const handleElementType = (element: GenericChannel) => {
+    switch (element.type) {
+      case 'user':
+        return upsertDirectMessage([(element.resource as UserType).id || '', currentUserId]);
+      case 'workspace':
+        return joinChannel(element.resource as ChannelResource);
+      case 'direct':
+        return upsertDirectMessage((element.resource as ChannelResource).data.members || []);
+    }
+  };
+
+  const upsertDirectMessage = async (userIds: string[]): Promise<void> => {
+    await ChannelsService.openDiscussion(userIds, companyId);
+    return ModalManager.closeAll();
+  };
 
-  const isJoined = (workspaceChannel: WorkspaceChannel) => {
-    return mine.some(
-      channel =>
-        workspaceChannel.channelResource.id === channel.id && channel.data.user_member?.user_id,
-    );
+  const joinChannel = (channel: ChannelResource) => {
+    const collectionPath: string = `/channels/v1/companies/${channel.data.company_id}/workspaces/${channel.data.workspace_id}/channels/${channel.data.id}/members/`;
+    const channelMembersCollection = Collection.get(collectionPath, ChannelMemberResource);
+    const findMember = channelMembersCollection.find({ user_id: currentUserId });
+
+    if (!findMember.length) {
+      channelMembersCollection.insert(
+        new ChannelMemberResource({
+          channel_id: channel.data.id,
+          user_id: currentUserId,
+          type: 'member',
+        }),
+      );
+    }
+
+    ModalManager.closeAll();
+    return RouterServices.history.push(`/client/${channel.data.workspace_id}/c/${channel.data.id}`);
```

Comment:
Not for this PR, but don't we have a service to hide this type of code, like `SomeService.open(channel)`? If not it will be good